## Упражнения по библиотеке Numpy

In [1]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [2]:
ar = np.random.randint(0, 10, size=10)
print(ar)
print(np.vectorize(lambda x: -x if 3 < x < 8 else x)(ar))

[7 3 2 3 3 6 3 6 8 6]
[-7  3  2  3  3 -6  3 -6  8 -6]


**2.** Заменить максимальный элемент случайного массива на 0

In [3]:
ar = np.random.randint(0, 10, size=10)
print(ar)
ar[ar.argmax()] = 0
print(ar)

[4 3 0 7 2 8 0 8 6 5]
[4 3 0 7 2 0 0 8 6 5]


**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [4]:
import itertools

ar = np.random.randint(0, 6, size=(3, 2))
print(ar)
np.array(list(itertools.product(*ar)))

[[5 0]
 [4 1]
 [0 3]]


array([[5, 4, 0],
       [5, 4, 3],
       [5, 1, 0],
       [5, 1, 3],
       [0, 4, 0],
       [0, 4, 3],
       [0, 1, 0],
       [0, 1, 3]])

**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [5]:
A = np.random.randint(0, 4, size=(8, 3))
B = np.random.randint(0, 4, size=(2, 2))
A, B

(array([[0, 3, 3],
        [1, 1, 1],
        [2, 2, 3],
        [3, 0, 2],
        [1, 0, 1],
        [1, 0, 0],
        [1, 1, 0],
        [1, 3, 2]]),
 array([[1, 3],
        [2, 0]]))

In [6]:
np.any(np.isin(A, B[0]), axis=1) & np.any(np.isin(A, B[1]), axis=1)


array([ True, False,  True,  True,  True,  True,  True,  True])

**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [7]:
A = np.random.randint(0, 2, size=(10, 3))
A

array([[1, 1, 1],
       [1, 1, 0],
       [0, 1, 1],
       [0, 0, 0],
       [1, 0, 0],
       [0, 1, 1],
       [0, 1, 1],
       [1, 1, 1],
       [1, 1, 0],
       [1, 0, 1]])

In [8]:
A[np.apply_along_axis(lambda x: np.unique(x).size != 1, 1, A)]

array([[1, 1, 0],
       [0, 1, 1],
       [1, 0, 0],
       [0, 1, 1],
       [0, 1, 1],
       [1, 1, 0],
       [1, 0, 1]])

**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [9]:
A = np.random.randint(0, 3, size=(10, 3))
A

array([[1, 1, 2],
       [0, 2, 1],
       [2, 0, 1],
       [1, 2, 1],
       [1, 2, 2],
       [1, 2, 0],
       [2, 1, 2],
       [1, 1, 0],
       [0, 0, 1],
       [1, 1, 0]])

In [10]:
indexes = np.full(A.shape[0], True)
for i in range(A.shape[0]):
    for j in range(i+1, A.shape[0]):
        if not indexes[j]:
            break
        if np.array_equal(A[i], A[j]):
            indexes[i] = False
            indexes[j] = False
A[indexes]

array([[1, 1, 2],
       [0, 2, 1],
       [2, 0, 1],
       [1, 2, 1],
       [1, 2, 2],
       [1, 2, 0],
       [2, 1, 2],
       [0, 0, 1]])

______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [11]:
def python_solution1(array):
    p = 1
    for i in range(min(len(array), len(array[0]))):
        if array[i][i] != 0:
            p *= array[i][i]
    return p

def numpy_solution1(a):
    diag = np.diag(a)
    return np.product(diag[diag != 0])

a = [[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]
python_solution1(a), numpy_solution1(np.array(a))

(3, 3)

* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [12]:
def python_solution2(x, y):
    from collections import Counter
    return Counter(x) == Counter(y)

def numpy_solution2(x, y):
    xcounts = np.dstack(np.unique(x, return_counts=True))
    ycounts = np.dstack(np.unique(y, return_counts=True))
    return np.setxor1d(xcounts, ycounts).size == 0

x = [1, 2, 2, 4]
y = [4, 2, 1, 2]

python_solution2(x, y), numpy_solution2(np.array(x), np.array(y))

(True, True)

* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [13]:
def python_solution3(x):
    mx = float('-inf')
    for i in range(1, len(x)):
        if x[i-1] == 0:
            mx = max(mx, x[i])
    return mx

def numpy_solution3(x):
    mask = x == 0
    mask[-1] = False
    return x[np.roll(mask, 1)].max()

x = [6, 2, 0, 3, 0, 0, 5, 7, 0]
python_solution3(x), numpy_solution3(np.array(x))

(5, 5)

* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [14]:
def python_solution4(x):
    values = []
    counts = []
    value = x[0]
    count = 1
    for i in range(1, len(x)):
        if x[i] == value:
            count += 1
        else:
            values.append(value)
            counts.append(count)
            value = x[i]
            count = 1
    values.append(value)
    counts.append(count)
    return (values, counts)

def numpy_solution4(x):
    indexes = np.nonzero(np.diff(x) != 0)[0]
    indexes = np.concatenate(([0], indexes + 1))
    counts = np.concatenate((np.diff(indexes), [1]))
    return (x[indexes], counts)
    
x = [2, 2, 2, 3, 3, 3, 5]
python_solution4(x), numpy_solution4(np.array(x))

(([2, 3, 5], [3, 3, 1]), (array([2, 3, 5]), array([3, 3, 1])))

* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [15]:
def python_solution5(X, Y):
    n = len(X)
    m = len(Y)
    d = len(X[0])
    distances = [[0] * m for i in range(n)]
    for i in range(n):
        for j in range(m):
            for k in range(d):
                distances[i][j] += (X[i][k] - Y[j][k])**2
            distances[i][j] = distances[i][j] ** 0.5
    return distances

def numpy_solution5(X, Y):
    xl = np.sum(X**2, axis=1, keepdims=True)
    yl = np.sum(Y**2, axis=1)
    return np.sqrt(xl + yl + (X @ Y.T)*(-2))

X = [[1, 2, 3], [0, 0, 0], [0, 1, 0]]
Y = [[1, 2, 3], [0, 1, 1], [0, 1, 1]]

print(*python_solution5(X, Y), sep='\n')
print()
print(*numpy_solution5(np.array(X), np.array(Y)), sep='\n')


[0.0, 2.449489742783178, 2.449489742783178]
[3.7416573867739413, 1.4142135623730951, 1.4142135623730951]
[3.3166247903554, 1.0, 1.0]

[0.         2.44948974 2.44948974]
[3.74165739 1.41421356 1.41421356]
[3.31662479 1.         1.        ]


In [16]:
import time

from scipy.spatial.distance import cdist

d = 100
n = 500
m = 600

t1 = 0
t2 = 0

steps = 50
for _ in range(steps):
    X = np.random.rand(n, d)
    Y = np.random.rand(m, d)

    start_time = time.perf_counter()
    m1 = numpy_solution5(X, Y)
    end_time = time.perf_counter()
    t1 += end_time - start_time

    start_time = time.perf_counter()
    m2 = cdist(X, Y, metric='euclidean')
    end_time = time.perf_counter()
    t2 += end_time - start_time

    assert np.allclose(m1, m2)

print(t1 / steps, t2 / steps)

0.05185496287998831 0.14671975089997433


_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [17]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [18]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [19]:
average_calories = np.average(calorie_stats)
average_calories

106.88311688311688

3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [20]:
calorie_stats_sorted = np.sort(calorie_stats)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [21]:
median_calories = np.median(calorie_stats_sorted)
median_calories

110.0

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [22]:
percentiles = np.percentile(calorie_stats_sorted, np.linspace(0, 50, 50)) 
# 0-50th percentile
percentiles

array([ 50.        ,  50.        ,  50.        ,  56.53061224,
        70.        ,  70.        ,  76.53061224,  84.28571429,
        90.        ,  90.        ,  90.        ,  90.        ,
        90.        ,  90.        ,  90.        ,  90.        ,
        94.08163265, 100.        , 100.        , 100.        ,
       100.        , 100.        , 100.        , 100.        ,
       100.        , 100.        , 100.        , 100.        ,
       100.        , 100.        , 100.        , 100.        ,
       100.        , 100.        , 100.        , 100.        ,
       100.        , 100.        , 104.69387755, 110.        ,
       110.        , 110.        , 110.        , 110.        ,
       110.        , 110.        , 110.        , 110.        ,
       110.        , 110.        ])

In [23]:
nth_percentile = np.argmax(percentiles > 60)
nth_percentile

4

6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [24]:
more_calories = np.sum(calorie_stats > 60) / calorie_stats.size * 100
more_calories

96.1038961038961

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [25]:
calorie_std = np.std(calorie_stats)
calorie_std

19.35718533390827

-- Стандартное отклонение не слишком маленькое, мы не можем обобщить что большинство злаков содержит около 100 калорий. Большое стандартное отклонение показывае что наши данные разняться

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Проанализировав данные о калорийности батончиков конкурентов можно сделать вывод что 
Батончики CrunchieMunchies имеют хороший показатель калорий в сравнении с конкурентами.
Хоть на рынке и есть компании с лучшими показателями чем у нас, наша продукция на высоком уровне.
При маркетинге можно показать насколько мы лучше чем другие:
1) "Батончики CrunchieMunchies менее калорийны чем 96% батончиков других брендов"
2) Наши батончики на 45% менее калорийны чем обычные*

*наши - 60ккал, средний показатель - 106ккал (43% improvement)

Такой анализ данных помогает понять конкурентоспособность свей продукции и обозначить свои сильные стороны